File: medium/go/2352-equal-row-and-column-pairs.go
package main

import (
	"strconv"
	"strings"
)

func equalPairs(grid [][]int)int {
	rowMap := make(map[string]int)
	n := len(grid)

	for i := 0; i < n; i++ {
		var rowStr  strings.Builder
		for j := 0; j < n; j++ {
			rowStr.WriteString("|")
			rowStr.WriteString(strconv.Itoa(grid[i][j]))
			rowStr.WriteString("|")
		}
		occur, ok := rowMap[rowStr.String()]
		if ok {
			rowMap[rowStr.String()] = occur+1
		} else {
			rowMap[rowStr.String()] = 1
		}
	}

	equalPairs := 0

	for i := 0; i < n; i++ {
		var colStr strings.Builder
		for j := 0; j < n; j++ {
			colStr.WriteString("|")
			colStr.WriteString(strconv.Itoa(grid[j][i]))
			colStr.WriteString("|")
		}

		occur, ok := rowMap[colStr.String()]

		if ok {
			equalPairs += occur
		}
	}

	return equalPairs 
}

File: medium/javascript/2352-equal-row-and-column-pairs.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var equalPairs = function (grid) {
  const rowMap = new Map();
  const n = grid.length;

  for (let i = 0; i < n; i++) {
    let rowStr = "";
    for (let j = 0; j < n; j++) {
      rowStr = rowStr.concat(`|${grid[i][j]}|`);
    }

    const occurrences = rowMap.get(rowStr);
    if (occurrences) {
      rowMap.set(rowStr, occurrences + 1);
    } else {
      rowMap.set(rowStr, 1);
    }
  }

  let equalPairs = 0;

  for (let i = 0; i < n; i++) {
    let colStr = "";
    for (let j = 0; j < n; j++) {
      colStr = colStr.concat(`|${grid[j][i]}|`);
    }
    const occurrences = rowMap.get(colStr);
    if (occurrences) {
      equalPairs += occurrences;
    }
  }

  return equalPairs;
};
